net/node/infoUpdate.go

```diff
@@ -122,58 +122,55 @@ func getNodeAddr(n *node) NodeAddr {
 	return addr
 }
 
-func (node *node) reconnect(peer *node) error {
-	isTls := config.Parameters.IsTLS
-	addr := getNodeAddr(peer)
-	var ip net.IP
-	ip = addr.IpAddr[:]
-	nodeAddr := ip.To16().String() + ":" + strconv.Itoa(int(addr.Port))
-	log.Info("try to reconnect peer, peer addr is ", nodeAddr)
-	var conn net.Conn
-	var err error
-	if isTls {
-		conn, err = TLSDial(nodeAddr)
-		if err != nil {
-			return nil
-		}
-	} else {
-		conn, err = NonTLSDial(nodeAddr)
-		if err != nil {
-			return nil
+func (node *node) reconnect() {
+	node.RetryConnAddrs.Lock()
+	defer node.RetryConnAddrs.Unlock()
+	lst := make(map[string]int)
+	for addr := range node.RetryAddrs {
+		node.RetryAddrs[addr] = node.RetryAddrs[addr] + 1
+		rand.Seed(time.Now().UnixNano())
+		log.Trace("Try to reconnect peer, peer addr is ", addr)
+		<-time.After(time.Duration(rand.Intn(CONNMAXBACK)) * time.Second)
+		log.Trace("Back off time`s up, start connect node")
+		node.Connect(addr)
+		if node.RetryAddrs[addr] < MAXRETRYCOUNT {
+			lst[addr] = node.RetryAddrs[addr]
 		}
 	}
-	t := time.Now()
-	peer.UpdateRXTime(t)
-	peer.tryTimes = 0
-	peer.conn = conn
-	peer.addr, err = parseIPaddr(conn.RemoteAddr().String())
-	peer.local = node
-
-	log.Info(fmt.Sprintf("Reconnect node %s connect with %s with %s",
-		conn.LocalAddr().String(), conn.RemoteAddr().String(),
-		conn.RemoteAddr().Network()))
-	go peer.rx()
-
-	if node.GetID() > peer.GetID() {
-		peer.SetState(HAND)
-		buf, _ := NewVersion(node)
-		peer.Tx(buf)
-	}
+	node.RetryAddrs = lst
 
-	return nil
 }
 
-func (node *node) TryConnect() {
-	node.nbrNodes.RLock()
-	defer node.nbrNodes.RUnlock()
-	for _, n := range node.nbrNodes.List {
-		if n.GetState() == INACTIVITY && n.tryTimes < 3 {
-			//try to connect
-			n.tryTimes++
-			node.reconnect(n)
-		}
+func (n *node) TryConnect() {
+	if n.fetchRetryNodeFromNeiborList() > 0 {
+		n.reconnect()
 	}
+}
 
+func (n *node) fetchRetryNodeFromNeiborList() int {
+	n.nbrNodes.RLock()
+	defer n.nbrNodes.RUnlock()
+	var ip net.IP
+	neibornodes := make(map[uint64]*node)
+	for _, tn := range n.nbrNodes.List {
+		addr := getNodeAddr(tn)
+		ip = addr.IpAddr[:]
+		nodeAddr := ip.To16().String() + ":" + strconv.Itoa(int(addr.Port))
+		if tn.GetState() == INACTIVITY {
+			//add addr to retry list
+			n.AddInRetryList(nodeAddr)
+			//close legacy node
+			if tn.conn != nil {
+				tn.CloseConn()
+			}
+		} else {
+			//add others to tmp node map
+			n.RemoveFromRetryList(nodeAddr)
+			neibornodes[tn.GetID()] = tn
+		}
+	}
+	n.nbrNodes.List = neibornodes
```

Comment:
notice that it defines two struct with a RWMutex.
```golang
type RetryConnAddrs struct {
	sync.RWMutex
	RetryAddrs map[string]int
 }
type ConnectingNodes struct {
	sync.RWMutex
	ConnectingAddrs []string
}
```
and move the addr in one list to another not in an atomic step ,  which will cause concurrent consistency problem。maybe need managed with the same Lock like the following struct :
```golang
type connAddrs struct {
	sync.RWMutex
	RetryAddrs map[string]int
        ConnectingAddrs []string
 }
```